packages/salesforcedx-vscode-core/src/conflict/metadataCacheService.ts

```diff
@@ -0,0 +1,285 @@
+      const propsFile = this.saveProperties(properties);
+      const cacheCommon = this.findLongestCommonDir(components, this.cachePath);
+
+      const projCommon = this.findLongestCommonDir(
+        this.sourceComponents.getSourceComponents().toArray(),
+        this.projectPath
+      );
+
+      const isPathDirectory =
+        fs.existsSync(this.componentPath) &&
+        fs.lstatSync(this.componentPath).isDirectory();
+
+      return {
+        selectedPath: this.componentPath,
+        selectedIsDirectory: isPathDirectory,
+
+        cache: {
+          baseDirectory: this.cachePath,
+          commonRoot: cacheCommon,
+          components
+        },
+        cachePropPath: propsFile,
+
+        project: {
+          baseDirectory: this.projectPath,
+          commonRoot: projCommon,
+          components: this.sourceComponents.getSourceComponents().toArray()
+        }
+      };
+    }
+  }
+
+  private extractResults(
+    result: RetrieveResult | SourceRetrieveResult
+  ): { components: SourceComponent[]; properties: FileProperties[] } {
+    let components: SourceComponent[] = [];
+    const properties: FileProperties[] = [];
+
+    if (result instanceof RetrieveResult) {
+      if (Array.isArray(result.response.fileProperties)) {
+        properties.push(...result.response.fileProperties);
+      } else {
+        properties.push(result.response.fileProperties);
+      }
+      components = result.components.getSourceComponents().toArray();
+    } else {
+      result.successes?.forEach(s => {
+        if (s.properties) {
+          properties.push(s.properties);
+        }
+        components.push(s.component);
+      });
+    }
+    return { components, properties };
+  }
+
+  private findLongestCommonDir(
+    comps: SourceComponent[],
+    baseDir: string
+  ): string {
+    if (comps.length === 0) {
+      return baseDir;
+    }
+    if (comps.length === 1) {
+      return this.getRelativePath(comps[0], baseDir);
+    }
+
+    const allPaths = comps.map(c => this.getRelativePath(c, baseDir));
+    const baseline = allPaths[0];
+    let shortest = baseline.length;
+
+    for (let whichPath = 1; whichPath < allPaths.length; whichPath++) {
+      const sample = allPaths[whichPath];
+      shortest = Math.min(shortest, sample.length);
+
+      for (let comparePos = 0; comparePos < shortest; comparePos++) {
+        if (baseline[comparePos] !== sample[comparePos]) {
+          shortest = comparePos;
+          break;
+        }
+      }
+    }
+
+    const dir = baseline.substring(0, shortest);
+    return dir.endsWith(path.sep) ? dir.slice(0, -path.sep.length) : dir;
+  }
+
+  private saveProperties(properties: FileProperties[]): string {
+    const props = {
+      componentPath: this.componentPath,
+      fileProperties: properties
+    };
+    const propDir = this.getPropsPath();
+    const propsFile = path.join(propDir, MetadataCacheService.PROPERTIES_FILE);
+
+    fs.mkdirSync(propDir);
```

Comment:
Will this error if propsDir already exists?

Reply:
@brpowell The entire cache directory is cleared (removed) before the retrieve, so there shouldn't be a file there.

Reply:
I see. Is the `MetadataCacheExecutor` the main consumer of the service?

Reply:
Yes, that covers the use cases that we are working on right now. I added the loadCache() function for convenience outside of an executor, but there is no use case for it at the moment.